doc: refactor metaprogramming section

ihnorton opened the pull request:
This is a fairly substantial reworking of parts of the Metaprogramming section of the manual, inspired by [this email chain](https://groups.google.com/d/msg/julia-users/bLDuSI1MkVA/vPOo7pPrwO0J) (among many others).

One of the main goals here is to introduce and work with directly-constructed Expr objects first, and separately from quoted expressions. This seems to be the way that many users end up using macros anyway, and I hope that decoupling expression structure will make both expressions and quoting easier to understand.

I've also tried to explicitly ask and answer the question "why are macros necessary", but that section may need some more work.

This PR does make the chapter longer, but hopefully breaks some of it into explicit and digestible chunks. I'll try to figure out how to add an internal TOC to the top of the page. Any comments very much appreciated.

cc: @tlycken @ismael-vc @jiahao

ihnorton committed:
doc: rewrite metaprog expression intro

[ci skip]

jiahao commented:
![i_love_dg_bbq_apron](https://cloud.githubusercontent.com/assets/1732/5604842/ba2d2506-93a5-11e4-95d4-a7a88134cde0.jpg)

([Photo credit](http://www.cafepress.com/+i_love_dg_bbq_apron,90555200))

ihnorton commented:
Oh man. That's low. This is my reward for working on documentation?!? :crying_cat_face:

jakebolewski commented:
Documentation, the housework of open source projects...

jakebolewski commented on doc/manual/metaprogramming.rst:
```diff
@@ -9,106 +9,120 @@
-      typ
-    end
+extra build steps, and also allows true Lisp-style macros operating at
+the level of abstract syntax. In contrast, preprocessor "macro" systems,
```
abstract syntax trees (with a link explaining the term)?

tlycken replied:
There was also a question on the mailing list about what "Lisp-style" means in this context. I don't know - and honestly, I don't think being "Lisp-style" is the most important trait of our macros - so I vote for skipping this description entirely. If not, it should probably be explained with a link to somewhere.

ihnorton replied:
@jakebolewski done

@ tlycken I think "Lisp-style" is descriptive for people who understand what it means and reasonably neutral otherwise (i.e. "keep reading to find out")

jakebolewski commented on doc/manual/metaprogramming.rst:
```diff
@@ -9,106 +9,120 @@
+    julia> ex3 = parse("(4 + 4) / 2")
+    :((4 + 4) / 2)
+
+    julia> Base.Meta.show_sexpr(ex3)
```
you don't need `Base` here.

timholy replied:
Since the topic is "display," one alternative phrasing might be something like:

Another way to view expressions is with `Meta.show_sexpr`, which displays the `S-expression <http://en.wikipedia.org/wiki/S-expression>`_ form of a given :obj:`Expr`, which may look very familiar to users of Lisp. Here's an example illustrating the display on a nested :obj:`Expr`:

ihnorton replied:
done and done

jakebolewski commented on doc/manual/metaprogramming.rst:
```diff
@@ -134,19 +162,88 @@ ambiguity in parsing.:
+are symbols, ``*(b,c)`` is a subexpression, and ``1`` is a literal
+64-bit signed integer.
+
+There is a second form of quoting for multiple expressions:
```
"syntactic form"? 

Maybe it would be good to say here that `:(a + b)` and `quote a + b end` are not equivalent at the expression level.  Also, `quote` blocks introduce `QuoteNode`'s into the AST which is a bit confusing to newcomers.  That is a bit of the internal IR leaking through but you quickly run into it if you work with the AST of parsed julia source code.

ihnorton replied:
done and done

jakebolewski commented on doc/manual/metaprogramming.rst:
```diff
@@ -134,19 +162,88 @@ ambiguity in parsing.:
+
+    julia> ex = :($a + b)
+    :(1 + b)
+
```
I think this example might be a bit too simple, I feel this is one of the topics people get the most confused by so it might be worth expanding a bit.

Some simple examples I have seen people get tripped up by:
```julia
# wrong
a = 1
ex = :($a in $((1,2,3)))

# right
a = 1
ex = :($a in $(Expr(:tuple, 1,2,3)))
```

I think people get confused because numbers are literal values in the AST so they think that all literal values can be interpolated into the expression.

Another case where this comes up is interpolating symbols which might deserve a special mention.
```julia
# ex
:a in (:a, :b)

#wrong
:($:a in $(Expr(:tuple, :a, :b)))

# right
:($(:(:a)) in $(Expr(:tuple, :(:a), :(:b)))))
```
Something about nested interpolations would also be a good "advanced" example but I cannot think of one off the top of my head.

ihnorton replied:
Thanks, those are good examples - will add.

timholy replied:
I agree that interpolation is one of the hardest topics to get right. +1 for Jake's symbol-interpolation example. Here's a likely gotcha, though:
```
julia> ex = :($(:(:a)))
:(:a)                                                                                                                                                                                                                                                                          
                                                                                                                                                                                                                                                                               
julia> ex1 = :($(Expr(:quote, :a)))
:(:a)                                                                                                                                                                                                                                                                          
                                                                                                                                                                                                                                                                               
julia> ex == ex1
false                                                                                                                                                                                                                                                                          

julia> Meta.show_sexpr(ex)
(:quote, #QuoteNode
  :a
)
julia> Meta.show_sexpr(ex1)
(:quote, :a)
```
For many purposes these can be used interchangeably, of course.

Being a `cartesian.jl/multidimensional.jl` kind of guy, I also find this to be an example worth showing (it's a common pattern in `base/`):
```julia
julia> args = [symbol(string("i_",j)) for j = 1:3]
3-element Array{Symbol,1}:
 :i_1
 :i_2
 :i_3

julia> ex = :(foo($(args...)))
:(foo(i_1,i_2,i_3))
```

ihnorton replied:
Alright, I added several of Jakes examples -- simplified, maybe, by interpolating quoted expressions. Instead of adding more content here, I think we should make a FAQ entry with advanced interpolation examples and link to it from here.

jakebolewski commented on doc/manual/metaprogramming.rst:
```diff
@@ -226,89 +322,112 @@ the important distinction between the way ``a`` and ``b`` are used:
+Macros are the analogue of functions for expression generation at
+compile time. Just as a function maps a tuple of argument values to a
+return value, a macro maps a tuple of argument *expressions* to a returned
+*expression*.
```
Macros can also take symbols as arguments.

timholy replied:
And literal numbers. But in a sense calling these expressions makes sense.

ihnorton replied:
Addressed

jakebolewski commented on doc/manual/metaprogramming.rst:
```diff
@@ -538,6 +660,57 @@ is occasionally quite handy.
+      end)
+    end
+
+In this manner, Julia acts as its own preprocessor, and allows code
```
link for "preprocessor" (for non CS people)?

jakebolewski commented on doc/manual/metaprogramming.rst:
```diff
@@ -538,6 +660,57 @@ is occasionally quite handy.
+
+    julia> $a + b
+    ERROR: unsupported or misplaced expression $
+
```
I think this section should end with a warning about code generation, specifically generating duplicate methods bodies of redundant code over many argument types.  

```julia

foo(arg1::Union(Type1, Type2), arg2::Union(Type1, Type2)) = ...

#vs

for (t1, t2) in [(:Type1, :Type1), (:Type1, :Type2), :(:Type2, :Type1), (:Type2, :Type2)]
    @eval foo(arg1::$t1, arg2::$t2) = ...
end
``` 

code generation for this purpose seems to get abused a lot.

timholy replied:
@jakebolewski, care to clarify what you mean by this remark? There are places where you need the second in order to get the dispatch result you want (`Union`s are penalized when it comes to method matching).

ihnorton replied:
I'll leave this one for future edits :)

jakebolewski commented:
:+1: I've added some comments where I have seen confusion.  Take them as you may as they are just suggestions, this is a big improvement.

ihnorton commented:
Thanks for all the comments! I'll incorporate them tomorrow (can't stand to stare at this any more today).

tlycken commented on doc/manual/metaprogramming.rst:
```diff
@@ -9,106 +9,120 @@
 support. Like Lisp, Julia represents its own code as a data structure of
 the language itself.
+Because all data types and code in Julia are represented by Julia data structures,
+powerful reflection capabilities are available to explore the internals of a
```
"reflection" might not be a well-known term among non-CS people. Link to e.g. wikipedia?

ihnorton replied:
done

tlycken commented on doc/manual/metaprogramming.rst:
```diff
@@ -9,106 +9,120 @@
+Because all data types and code in Julia are represented by Julia data structures,
+powerful reflection capabilities are available to explore the internals of a
+program and its types just like any other data.
 Since code is represented by objects that can be created and manipulated
```
It seems to me that this sentence says almost exactly the same thing as the (new) one before it. Can they be re-worked into one, or maybe even just scratch one of them?

ihnorton replied:
Moved to the end. But, on this point, I think repetition is quite a virtue :hammer:.

tlycken commented on doc/manual/metaprogramming.rst:
```diff
@@ -9,106 +9,120 @@
+
+.. doctest::
+
+    julia> ex2 = Expr(:call, :+, 1, 1)
```
I don't think `Symbol`s have been introduced earlier in the manual, so somewhere around here might be a good time. Perhaps something as simple as

> `Symbol`s are a special kind of string literal, that can only take values that are valid identifiers *(is that true?)*, denoted by prefixing the string literal with a `:` instead of wrapping it in citation marks. You can also construct symbols from run-time values using `symbol(your_string)`.

I don't know if it's a good idea to interrupt this example, though, so we might want to splice this in somewhere a little earlier.

ihnorton replied:
Added a note to "see discussion of symbols below". This is one of those things that I hope an active reader will notice and look for an answer -- maybe that's a bad idea though. We could move the discussion of symbols to the very beginning instead, but IMHO that would be a really dry introduction.

timholy commented on doc/manual/metaprogramming.rst:
```diff
@@ -134,19 +162,88 @@ ambiguity in parsing.:
+
+Direct construction of :obj:`Expr` objects with value arguments is
+powerful, but :obj:`Expr` constructors can be tedious compared to "normal"
+Julia syntax. Instead, Julia allows "splicing" or interpolation of expression
```
"Instead" -> "As an alternative"? (Sometimes I've found it actually easier to use `Expr` directly.)

ihnorton replied:
done

timholy commented on doc/manual/metaprogramming.rst:
```diff
@@ -226,89 +322,112 @@ the important distinction between the way ``a`` and ``b`` are used:
+	   end
 
-.. _man-macros:
+The "return value" of this macro is the *interpolated* expression; that is,
```
Might it make sense to first show what happens when you use it?
```julia
julia> @sayhello "human"
Hello, human!
```

ihnorton replied:
Will add

timholy commented on doc/manual/metaprogramming.rst:
```diff
@@ -226,89 +322,112 @@ the important distinction between the way ``a`` and ``b`` are used:
+
+Macros are useful because they allow the programmer to
+transform generated code into expanded expressions
+*before* the full program is compiled and executed.
```
Perhaps add "Macros execute at the time code is parsed, not at the time it is run."

timholy commented on doc/manual/metaprogramming.rst:
```diff
@@ -226,89 +322,112 @@ the important distinction between the way ``a`` and ``b`` are used:
+
+.. doctest::
+
+    julia> eval(macroexpand( quote @sayhello("human") end ))
```
Amplifying my comment below, it's a little strange to have this here before the "Macro invocation" section and the introduction of `@`.

ihnorton replied:
To me this is a chicken/egg thing. It seems equally strange to say "`@foo` invokes the expanded result of `macro foo`..." without having first introduced expansion. I think this distinction really trips people up, and I'm trying to find a way to clearly delineate and emphasize the idea that `@foo` is mostly about producing an expression (at least for the sake of effectively writing macros).

(also, hopefully people will notice something funny/new and keep reading for the explanation)

timholy replied:
I'm fine with whatever choice you make.

timholy commented:
:100: Big-time thanks for tackling this!

elextr commented:
Its hard to add to the comments already made so just add thanks and well done.  Now who's for `stagedfunction`s :-P

tlycken commented on doc/manual/metaprogramming.rst:
```diff
@@ -134,19 +162,88 @@ ambiguity in parsing.:
+Quoting
+~~~~~~~
+
+There is special syntax to create expression objects without using the
```
In your opening paragraph about symbols, you state that 

> The `:` character has two syntactic purposes in Julia. The first form creates a [...]

It would be nice to refer back to this statement, maybe re-wording this as

> The second syntactic purpose of the  `:` character is to construct expression objects without using the explicit `Expr` constructor; this is sometimes referred to as quoting *quoting*."

ihnorton replied:
Used your wording (much better) and re-worded the following sentences.

tlycken commented on doc/manual/metaprogramming.rst:
```diff
@@ -134,19 +162,88 @@ ambiguity in parsing.:
+	x + y
+    end
+
+    julia> typeof(ex)
```
```
julia> Meta.show_sexpr(ex)
(:block,
  (:line, 2, :none),
  (:(=), :x, 1),
  :( # line 3:),
  (:(=), :y, 2),
  :( # line 4:),
  (:call, :+, :x, :y)
)
```

might also be interesting here, to show that `quote ... end` simply wraps the quoted expressions in a block (expression).

ihnorton replied:
I'm hoping people will try that themselves.

tlycken commented on doc/manual/metaprogramming.rst:
```diff
@@ -226,89 +322,112 @@ the important distinction between the way ``a`` and ``b`` are used:
+Macros are invoked with the following general syntax::
 
     @name expr1 expr2 ...
     @name(expr1, expr2, ...)
```
I seem to recall that this syntax was discouraged, but I can't seem to find the julia-users threads where I read this... If I do remember it correctly, though, being satisfied with only one macro call syntax (with space, no parens), will allow for cutting out almost the entire next paragraph; shortening this page can't be a bad thing.

timholy replied:
It's actually that `f (x)` (with a space between `f` and the parenthesis) is going to be disallowed, because of the issue described in the next paragraph.

I know what you mean about the length of the page, but since this is a "gotcha" I tend to think it needs to be described.

tlycken replied:
Ah, yes. That's it.

I wholeheartedly agree that gotcha's like this one need to be described; however, if we teach (especially) new users that `macros are called with `@name expr1, exp2, ...` we might eventually make it so that no-one ever uses `@macro(expr1, expr2, ...)`, and therefore never happens to try `@macro (expr1, expr2, ...)`. But as long as `@macro(expr1, expr2, ...)` is allowed, of course it should be in the manual.

timholy replied:
Even though I usually use the space version, the problem is that some macro calls are actually clearer (IMO) with parens. Read through some of the code in `multidimensional.jl` to see what I mean.

johnmyleswhite replied:
Agree that using parentheses can make macro calls a lot easier to read.

ihnorton replied:
I think both have their place; un-parenthesized form is really nice for things like `@inbounds` and wrappers (e.g. Cxx) where parentheses would be unsightly. Might as well let people know both are available.

jakebolewski replied:
also worth mentioning is that you can use the space sensitive macros anywhere by wrapping them in surrounding parens.

```julia
@cxx test->getfield() + 1 = 2 # Error
(@cxx test->getfield()) + 1 = 2
```

tlycken commented:
There, I've added some comments too. Most of them are minor stuff, so feel free to ignore anything you strongly disagree with.

And huge thanks for tackling this!

ihnorton commented:
I think I addressed most of the comments so far; thanks for all of them. It is, of course, a living document so this will certainly not be the final word :)

(ci canceled manually)

timholy commented:
I'd say merge it whenever you are ready. Documentation is never "finished."

nalimilan commented on doc/manual/metaprogramming.rst:
```diff
@@ -118,8 +137,22 @@ instead of an :obj:`Expr`:
+    julia> :foo == symbol("foo")
+    true
+
+    julia> symbol('\'')
```
Maybe introduce `symbol("'")` first, as I found it confusing that you switch to using `'` without any apparent reason. (Is the `'` notation really useful actually?)

elextr replied:
The ' ' pairing is a character (as mentioned in the text above), its documented [here](http://docs.julialang.org/en/latest/stdlib/punctuation/) which could be better, but its nothing to do with metaprogramming and its explanation IMO doesn't belong in this section.

nalimilan replied:
I'm not sure I was clear. I think introducing `symbol(::Char)` via `symbol('\'')` is not the clearest solution, as the reader may be confused by simultaneous introduction of two different meanings of `'`. Better start with `symbol("'")` first, since it's more similar to `symbol("foo")` which was introduced above; or replace `symbol('\'')` with `symbol('&')`. (Escaping with `\'` is not essential for metaprogramming, and this section is already dense enough.)

elextr replied:
Ok, agree that a normal unescaped character would be a better example.

ihnorton replied:
Removed the escaped version, doesn't seem necessary.

nalimilan commented on doc/manual/metaprogramming.rst:
```diff
@@ -134,19 +167,102 @@ ambiguity in parsing.:
+    julia> ex = :($a + b)  # interpolate the literal value of `a`
+    :(1 + b)
+
+    julia> ex = :(a in $:((1,2,3)) )
```
I don't understand this example. Maybe you could introduce it a bit more? (Same just below.)

ihnorton replied:
Added some clarification

nalimilan commented on doc/manual/metaprogramming.rst:
```diff
@@ -134,19 +167,102 @@ ambiguity in parsing.:
+Interpolating symbols into a nested expression requires enclosing each 
+symbol in an enclosing quote block:
+
+    julia> :( :a in $( :(:a + :b) ) )
```
Maybe put a series of `^^^^^` instead of just one at the beginning and one at the end.

nalimilan commented on doc/manual/metaprogramming.rst:
```diff
@@ -226,89 +341,113 @@ the important distinction between the way ``a`` and ``b`` are used:
-------
+    julia> ex = macroexpand( :(@sayhello("human")) )
+    :(println("Hello, ","human","!"))
+                        ^     ^
```
Same `^^^^^` here.

ihnorton replied:
done and done

dpsanders commented:
:100: Excellent rewrite!

On a first reading, my main comment is that it's maybe not clear that a macro behaves like a function that receives the code its given as an expression. (I know people often try to avoid saying that a macro is a function, but it's a function that acts on an expression, manipulates it, and returns a new expression.)

When I was trying to get to grips with this stuff, it was useful to define macros like

```
macro showarg(x)
       @show x
end
```

and 
```
macro showargs(x...)
       @show x
       @show x[1]
end
```

with examples like
```
julia> @showargs a b c
x = (:a,:b,:c)
x[1] = :a
ERROR: a not defined

julia> a = 3
3

julia> @showargs a b c
x = (:a,:b,:c)
x[1] = :a
3
```
so that I could understand in what form the macro was getting its arguments.

StefanKarpinski commented:
Thanks for the update, @ihnorton, this approach is a nice improvement :+1:

ihnorton committed:
doc: more metaprog edits

[ci skip]

ihnorton commented:
@dpsanders thanks for the comments, I added a modified version of your example (sans varargs), as well as another attempting to illustrate that a function called in an unquoted part of a `macro ... end` block is executed at expansion time.

Thanks for all of the suggestions, everyone.

timholy commented:
Thanks for taking this on! It's a really big improvement.

tkelman commented:
backported in becf36f5015c7248dba5a7793502c666c8a3ec7f
